src/runtime/manifest.ts

```diff
@@ -715,6 +719,32 @@ ${e.message}
     visitor.traverse(items);
   }
 
+  private static _disallowRecursiveSchemas(manifest: Manifest) {
+    for (const schema of Object.values(manifest.schemas)) {
+      const referenced: Set<string> = new Set();
+      const visit = (schema: Schema) => {
+        if (referenced.has(schema.name)) {
+          return; // already visited
+        }
+        // visit fields
+        for (const field of Object.values(schema.fields)) {
+          const entityType = field.getEntityType();
+          if (entityType == null) {
+            // ignore Primitive types
+            continue;
+          }
+          const fieldType = entityType.getEntitySchema();
+          referenced.add(fieldType.name);
+          visit(fieldType);
+        }
+      };
+      visit(schema);
+      if (referenced.has(schema.name)) {
+        throw new ManifestError(schema.location, `Unsupported recursive schema ${schema.name}`);
```

Comment:
Suggestion: you could say something like this instead: `Recursive schemas are unsupported, requires enabling flag recursiveSchemasAllowed: ${schema.name}.`